Upgrade LoggingTensor mode and add traceback collection

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103859
* __->__ #103734

Parts borrowed from: https://github.com/albanD/subclass_zoo/blob/main/logging_mode.py

soulitzer committed:
Upgrade LoggingTensor mode and add traceback collection

[ghstack-poisoned]

soulitzer committed:
Update on "Upgrade LoggingTensor mode and add traceback collection"




[ghstack-poisoned]

albanD commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -74,52 +94,100 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
 class LoggingTensorReentrant(LoggingTensor):
     context = torch.overrides.enable_reentrant_dispatch
 
+def _maybe_repr(a):
```
Why not `str(a)` ?

albanD commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -74,52 +94,100 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+                return
+            self.memo.pop(o, None)
+
+        weakref.finalize(t, del_memo)
```
Why do this by hand and no use `torch.utils.weak.WeakTensorKeyDictionary` for self.memo?

albanD commented on torch/testing/_internal/logging_tensor.py:
```diff
@@ -74,52 +94,100 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
 
 @contextlib.contextmanager
-def capture_logs(is_mode=False) -> Iterator[List[str]]:
+def capture_logs(is_mode=False, python_tb=False, script_tb=False, cpp_tb=False) -> Iterator[List[str]]:
```
I think the Mode version is the one I usually use, it would be great to add these args to capture_logs_with_logging_tensor_mode below

albanD reviewed (approved):
SGTM overall! Thanks!

soulitzer committed:
Update on "Upgrade LoggingTensor mode and add traceback collection"


Parts borrowed from: https://github.com/albanD/subclass_zoo/blob/main/logging_mode.py


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge -f "Unrelated failures"
